### PR TITLE
Make ForwardChainerUTest stable

### PR DESCRIPTION
**Pull request by vsbogd**
ForwardChainerUTest fails sporadically. To reproduce one can use the
following sequence of commands:
$ cd build/tests
$ while true; do \
    ctest -j6 --force-new-ctest-process --output-on-failure -R ForwardChainerUTest; \
  done
After some time test fails.

### CODE REVIEW

**Comment by linas** [comment]
This looks good to me; I'd merge it, except I don't want to step on your toes, @ngeiswei 

Incidentally, there is some bad indentation nearby... seems like inconsistent tab/space usage. I don't care which is used, but it should be consistent within the same file.

**Comment by ngeiswei** [comment]
Oh, I missed that one! @vsbogd just correct the indentation then I'll have a look at it, thanks.

**Comment by vsbogd** [comment]
I see that some files are formatted in mixed manner: some lines are formatted using tabs, some formatted using spaces, ForwardChainerUTest for instance. Would you mind if I will convert such file to uniform formatting (using spaces for instance) if I touched it?

**Comment by ngeiswei** [comment]
Geez, you're right. In principle I suppose you'd go with the predominant use, but here it seems to be 50 50. I would say convert with tabs (or even better smart tabs if you can) because the URE code and its test units are overwhelmingly indented with tabs.

I personally prefer tabs over whitespaces. In fact to have the best of both of worlds I use smart tabs (https://www.emacswiki.org/emacs/SmartTabs). But how you format your own code is up to you, as long as you follow the predominant formating of the existing files you edit.

**Comment by vsbogd** [comment]
Fixed

**Comment by linas** [comment]
OK. Silly me; I should have been clear. White-space fixes kind-of-ish should be a distinct pull request, separate from pull requests that perform functional modification.  The reason for this is to make code review less confusing.  It takes concentration and energy to read code changes, and its time consuming and painful to read diffs, only to realize that they alter whitespace.   Conersely, once one start seeing a lot of white-space changes, one tends to skim them. If there is a functional change buried in the middle, then its much too easy to skip it.

I am merging this now, just to be done with it. But, in the future, please make (major) white-space changes, and functional changes into distinct pull requests.

**Comment by vsbogd** [comment]
@linas , I am surprised github has no 'ignore whitespace' button, but it seems you can modify review URL to enable this mode https://github.com/opencog/atomspace/pull/1643/files?w=1

**Comment by linas** [comment]
Yeah, that's not the point. Its not about github or what magic buttons it might or might not have. Its about the actual changes being made to the code.

**Comment by vsbogd** [comment]
Probably it makes sense format all of code once and stick to single formatting style then.

**Comment by ngeiswei** [comment]
@vsbogd Why did you merge the master in the end? It probably wasn't necessary, unless your code only made sense with the last commits from the master. That's no big deal but it makes the github history slightly more convoluted than it needs to be.

File: tests/rule-engine/forwardchainer/ForwardChainerUTest.cxxtest [hunk 1]
```diff
@@ -38,14 +38,13 @@ public:
...
 		string source_dir = string(PROJECT_SOURCE_DIR),
-			test_dir = source_dir + "/tests",
-			test_ure_dir = test_dir + "/rule-engine",
-			test_forwardchainer_dir = test_ure_dir + "/forwardchainer",
-			ure_dir = source_dir + "/opencog/scm/opencog/rule-engine";
-		vector<string> load_paths = {source_dir, test_dir, test_ure_dir,
-		                             test_forwardchainer_dir, ure_dir};
-		for (string& p : load_paths)
-		{
+		       test_dir = source_dir + "/tests",
```
[the remark this answers is not in the document]
Comment (vsbogd) [reply tests/rule-engine/forwardchainer/ForwardChainerUTest.cxxtest]: I use Eclipse, it doesn't have exactly "smart tabs" but have "mixed tabs" with "use tabs only for leading indentation" option. I have checked with [Emacs SmartTabs page](https://www.emacswiki.org/emacs/SmartTabs) you sent and overlooked this difference. It cannot proceed few variable declarations after type correctly.

[the remark this answers is not in the document]
Comment (ngeiswei) [reply tests/rule-engine/forwardchainer/ForwardChainerUTest.cxxtest]: Oh, I see. Sorry I was misguided by the removed lines. It's good enough. Great to know Eclipse has something like smart tabs. Do you know a page explaining how to enable that functionality? I'd like to add it in the dev standard, next to the emacs smart tabs link.

**Comment by ngeiswei** [comment]
There's an emacs package called `dtrt-indent` that automatically selects the indentation style, unfortunately it doesn't support smart tabs, only whitespaces vs regular tabs. Adding smart tabs support to it is on my emacs TODO list.

File: tests/rule-engine/forwardchainer/ForwardChainerUTest.cxxtest (hunk 1 above)
[the remark this answers is not in the document]
Comment (vsbogd) [reply tests/rule-engine/forwardchainer/ForwardChainerUTest.cxxtest]: Unfortunately I cannot find the description even in Eclipse documentation ) I just experimented a bit to check that it works fine. Briefly:
Window->Preferences->C/C++->Code Style->Formatter->Active Profile->Edit..
In new window on "Indentation Style" tab: 
- set "Tab Policy" = Mixed
- set "Use tabs only for leading indentation"

**Comment by linas** [comment]
> Probably it makes sense format all of code once and stick to single formatting style then.

No, please do not do this, never-ever even joke about it, don't think about it ever again. The core problem with doing this is that it erases the file history, which makes debugging pointlessly painful, time-consuming and difficult.

**Comment by ngeiswei** [comment]
> > Probably it makes sense format all of code once and stick to single formatting style then.
> No, please do not do this, never-ever even joke about it, don't think about it ever again. The core problem with doing this is that it erases the file history, which makes debugging pointlessly painful, time-consuming and difficult.
OK, it's inconvenient, that's true, but I wouldn't go as far as saying it "makes debugging pointlessly painful". Nowadays IDEs easily allow to go back and forth between revisions of a file as you browse (Emacs certainly does).

**Comment by ngeiswei** [comment]
Anyway, I do agree that it's enough of an inconvenience to avoid it. Better make the IDE auto-detect and select the indentation style.

**Pull request merged**